core/src/main/java/nl/nn/adapterframework/compression/ZipWriter.java

```diff
@@ -15,284 +15,103 @@
-		closeEntry();
-		try {
-			if (isCloseOnExit()) {
-				zipoutput.close();
-			} else {
-				zipoutput.finish();
-			}
-		} catch (IOException e) {
-			throw new CompressionException("Cannot close ZipStream",e);
-		}
-	}
-
-	@Override
-	public Message writeItem(Message input, PipeLineSession session, ParameterValueList pvl, IZipWritingElement writingElement) throws CollectionException, TimeoutException {
+	public MessageZipEntry createPart(Message input, PipeLineSession session, ParameterValueList pvl) throws CollectionException {
 		String filename = ParameterValueList.getValue(pvl, PARAMETER_FILENAME, "");
-		String charset = writingElement.getCharset();
-		boolean completeFileHeader = writingElement.isCompleteFileHeader();
 		try {
 			Message contents = ParameterValueList.getValue(pvl, PARAMETER_CONTENTS, input);
-			if (StringUtils.isEmpty(filename) && contents!=input) {
+			if (StringUtils.isEmpty(filename) && contents != input) {
 				filename = input.asString();
 			}
-			if (completeFileHeader) {
-				writeEntryWithCompletedHeader(filename, contents, writingElement.isCloseInputstreamOnExit(), charset);
-			} else {
-				writeEntry(filename, contents, writingElement.isCloseInputstreamOnExit(), charset);
+			MessageZipEntry entry = new MessageZipEntry(contents, filename);
+			if (includeFileHeaders) {
+				entry.computeFileHeaders();
 			}
-			return contents==input ? Message.nullMessage() : input;
-		} catch (IOException | CompressionException e) {
+			return entry;
+		} catch (IOException e) {
 			throw new CollectionException("cannot write item", e);
 		}
 	}
 
 	@Override
-	public MessageOutputStream provideOutputStream(PipeLineSession session, ParameterValueList pvl, IZipWritingElement writingElement) throws CollectionException {
-		String filename = ParameterValueList.getValue(pvl, PARAMETER_FILENAME, "");
-		String charset = writingElement.getCharset();
-		boolean completeFileHeader = writingElement.isCompleteFileHeader();
-
-		if (completeFileHeader || StringUtils.isEmpty(filename)) {
-			return null;
-		}
-		Message contents = ParameterValueList.getValue(pvl, PARAMETER_CONTENTS, Message.nullMessage());
-		if (!Message.isNull(contents)) { // cannot provide OutputStream if contents is something else as the input message
-			return null;
-		}
-		try {
-			openEntry(filename);
-			OutputStream stream = StreamUtil.dontClose(getZipoutput());
-			return new MessageOutputStream(writingElement, stream, (IForwardTarget)null, charset) {
-
-				@Override
-				public void afterClose() throws Exception {
-					stream.flush();
-					closeEntry();
+	public Message build(List<MessageZipEntry> parts) throws IOException {
+		File collectorsTempFolder = FileUtils.getTempDirectory("collectors");
+		File tempFile = File.createTempFile("msg", ".zip", collectorsTempFolder);
+
+		try (FileOutputStream fos = new FileOutputStream(tempFile)) {
+			try (ZipOutputStream zipoutput = new ZipOutputStream(fos)) {
```

Comment:
volgens mij is dit in 1 try-with-resources regel te combineren door een `;` te zetten achter de eerste declaratie.